packages/bus-query-operation/lib/Bindings.ts

```diff
@@ -50,3 +52,195 @@ export function isBindings(maybeBindings: any): boolean {
 export function ensureBindings(maybeBindings: any): Bindings {
   return isBindings(maybeBindings) ? maybeBindings : Bindings(maybeBindings);
 }
+
+/**
+ * Materialize the given operation (recursively) with the given bindings.
+ * Essentially, all variables in the given operation will be replaced
+ * by the terms bound to the variables in the given bindings.
+ * @param {Operation} operation SPARQL algebra operation.
+ * @param {Bindings} bindings A bindings object.
+ * @param {boolean} strictTargetVariables If target variable bindings (such as on SELECT or BIND) should not be allowed.
+ * @return Algebra.Operation A new operation materialized with the given bindings.
+ */
+export function materializeOperation(operation: Algebra.Operation, bindings: Bindings,
+                                     strictTargetVariables: boolean = false): Algebra.Operation {
+  return Util.mapOperation(operation, {
+    path: (op: Algebra.Path, factory: Factory) => {
+      // Materialize variables in a path expression.
+      // The predicate expression will be recursed.
+      return {
+        recurse: false,
+        result: factory.createPath(
+          materializeTerm(op.subject, bindings),
+          op.predicate,
+          materializeTerm(op.object, bindings),
+          materializeTerm(op.graph, bindings),
+        ),
+      };
+    },
+    pattern: (op: Algebra.Pattern, factory: Factory) => {
+      // Materialize variables in the quad pattern.
+      return {
+        recurse: false,
+        result: factory.createPattern(
+          materializeTerm(op.subject, bindings),
+          materializeTerm(op.predicate, bindings),
+          materializeTerm(op.object, bindings),
+          materializeTerm(op.graph, bindings),
+        ),
+      };
+    },
+    extend: (op: Algebra.Extend, factory: Factory) => {
+      // Materialize an extend operation.
+      // If strictTargetVariables is true, we throw if the extension target variable is attempted to be bound.
+      // Otherwise, we bind the variable to the bound value directly.
+      if (bindings.has(termToString(op.variable))) {
+        if (strictTargetVariables) {
+          throw new Error(`Tried to bind variable ${termToString(op.variable)} in a BIND operator.`);
+        } else {
+          return {
+            recurse: true,
+            result: factory.createExtend(
+              op.input,
+              op.variable,
+              factory.createTermExpression(bindings.get(termToString(op.variable)))
+            ),
```

Comment:
Since every instance of variable op.variable would have been overwritten so far it would make more sense to replace the above with `result: op.input` right? Or am I missing a case where this could be an issue?

Reply:
Ah yes, you're right.